### PR TITLE
refactor(progress): rewrite raterun to make it more obvious

### PR DESCRIPTION
**Pull request by nvloff-f3**

### CODE REVIEW

File: internal/raterun/runner.go [hunk 1]
```diff
@@ -1,102 +1,117 @@
...
 
-// RunFunction function to be executed at specified rate
-type RunFunction func(rate time.Duration, t time.Time)
+// RunFunction function to be executed every period
+type RunFunction func(period time.Duration)
 
-type Rate struct {
-	Start time.Duration // when to start this rate
-	Rate  time.Duration // run in amount of Duration
+type Schedule struct {
```
Comment (janakerman-form3) [inline internal/raterun/runner.go:12]: nit: comments here explaining how the properties are used would help me.

// StartDelay is the delay that this schedule should start at.
// Frequency defines how often the provided RunFunction should be executed during this schedule.

File: internal/raterun/runner.go [hunk 2]
```diff
@@ -1,102 +1,117 @@
...
-	RestartRate()
-	// run starts running the function at rates provided to the constructor
-	Run()
-}
-
-// New creates a new runner which varies in time according to given rates
-func New(fn RunFunction, rates []Rate) (*Instance, error) {
-	if len(rates) == 0 {
-		return nil, errors.New("empty rates")
+func New(fn RunFunction, schedules []Schedule) (*Runner, error) {
```
Comment (janakerman-form3) [inline internal/raterun/runner.go:17]: nit: a comment here would help me, I think.

// New creates a new `Runner` that will execute `fn` as defined by the provided `schedules`.
// Each `Schedule` in `schedules` defines how often `fn` should be executed at any given period of time.

File: internal/run/test_runner.go [hunk 3]
```diff
@@ -160,9 +163,12 @@ func (r *Run) Do(ctx context.Context, s *scenarios.Scenarios) (*Result, error) {
 		}
 	}()
 
+	progressCtx, progressCancel := context.WithCancel(ctx)
```
Comment (janakerman-form3) [inline internal/run/test_runner.go:166]: What's your thinking RE favouring context as a form of cancellation as opposed to encapsulating it as a method of the progrss runner? I'm fairly indifferent in terms of which I prefer but curious as to your thinking.

Comment (nvloff-f3) [reply internal/run/test_runner.go]: added a stop func

**Review by janakerman-form3 — APPROVED** [review]

**Pull request merged**